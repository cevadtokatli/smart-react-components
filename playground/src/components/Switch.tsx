import { Switch, SwitchGroup } from '@smart-react-components/ui'
import React from 'react'

export const Palette = () => {
  const [active, setActive] = React.useState(() => ['primary'])

  return (
    <SwitchGroup active={active} setActive={setActive}>
      <Switch value="primary" label="Primary" />
      <Switch value="secondary" palette="secondary" label="Secondary" />
      <Switch value="success" palette="success" label="Success" />
      <Switch value="danger" palette="danger" label="Danger" />
      <Switch value="warning" palette="warning" label="Warning" />
      <Switch value="info" palette="info" label="Info" />
      <Switch value="dynamic" palette="dynamic" label="Dynamic" />
      <Switch value="!dynamic" palette="!dynamic" label="!Dynamic" />
    </SwitchGroup>
  )
}

export const Soft = () => {
  const [active, setActive] = React.useState(() => ['primary'])

  return (
    <SwitchGroup active={active} setActive={setActive}>
      <Switch value="primary" label="Primary" isOutline />
      <Switch value="secondary" palette="secondary" label="Secondary" isOutline />
      <Switch value="success" palette="success" label="Success" isOutline />
      <Switch value="danger" palette="danger" label="Danger" isOutline />
      <Switch value="warning" palette="warning" label="Warning" isOutline />
      <Switch value="info" palette="info" label="Info" isOutline />
      <Switch value="dynamic" palette="dynamic" label="Dynamic" isOutline />
      <Switch value="!dynamic" palette="!dynamic" label="!Dynamic" isOutline />
    </SwitchGroup>
  )
}

export const Outline = () => {
  const [active, setActive] = React.useState(() => ['primary'])

  return (
    <SwitchGroup active={active} setActive={setActive}>
      <Switch value="primary" label="Primary" isOutline />
      <Switch value="secondary" palette="secondary" label="Secondary" isOutline />
      <Switch value="success" palette="success" label="Success" isOutline />
      <Switch value="danger" palette="danger" label="Danger" isOutline />
      <Switch value="warning" palette="warning" label="Warning" isOutline />
      <Switch value="info" palette="info" label="Info" isOutline />
      <Switch value="dynamic" palette="dynamic" label="Dynamic" isOutline />
      <Switch value="!dynamic" palette="!dynamic" label="!Dynamic" isOutline />
    </SwitchGroup>
  )
}

export const OutlineSoft = () => {
  const [active, setActive] = React.useState(() => ['primary'])

  return (
    <SwitchGroup active={active} setActive={setActive}>
      <Switch value="primary" label="Primary" isOutline isSoft />
      <Switch value="secondary" palette="secondary" label="Secondary" isOutline isSoft />
      <Switch value="success" palette="success" label="Success" isOutline isSoft />
      <Switch value="danger" palette="danger" label="Danger" isOutline isSoft />
      <Switch value="warning" palette="warning" label="Warning" isOutline isSoft />
      <Switch value="info" palette="info" label="Info" isOutline isSoft />
      <Switch value="dynamic" palette="dynamic" label="Dynamic" isOutline isSoft />
      <Switch value="!dynamic" palette="!dynamic" label="!Dynamic" isOutline isSoft />
    </SwitchGroup>
  )
}

export const Size = () => {
  const [active, setActive] = React.useState(() => ['medium'])

  return (
    <SwitchGroup active={active} setActive={setActive}>
      <Switch value="small" size="small" label="Small" />
      <Switch value="medium" label="Medium" />
      <Switch value="large" size="large" label="Large" />
    </SwitchGroup>
  )
}

export const Block = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <SwitchGroup active={active} setActive={setActive} isBlock isRequired>
      <Switch value="item-1" label="Item 1" />
      <Switch value="item-2" label="Item 2" />
      <Switch value="item-3" label="Item 3" />
    </SwitchGroup>
  )
}
