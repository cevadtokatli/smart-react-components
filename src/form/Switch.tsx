import React from "react"
import styled from "styled-components"
import ElementProps from "../element/Props"
import {Size$StrictProps} from "../props/size$-strict-props"
import size from "../constants/form/switch-size"
import InputElement from "../element/Input"
import {SizeProps} from "../props/size-props"
import {JSXElementProps, JSXChild, Value, SetValue} from "../props"
import DV from "../default-value"
import useMultiple from "./UseMultiple"
import RadioLabel from "./RadioLabel"
import RadioSpan from "./RadioSpan"

const Foundation = styled.div``

const Slider = styled.div``

interface InputProps extends ElementProps, Size$StrictProps {
    type$: string
    fill$: boolean
    disabled?: boolean
}

const customKeys = {
    size$: (v, t) => `
        + ${Foundation} {
            width: ${size.foundation[v].width};
            height: ${size.foundation[v].height};

            > ${Slider} {
                width: ${size.slider[v].size};
                height: ${size.slider[v].size};
            }
        }

        &:checked + ${Foundation} > ${Slider} {
            left: ${size.slider[v].left};
        }

        ~ ${RadioSpan} {
            font-size: ${t.form.radioSpanFontSize[v]};
        }
    `
}

const Input = styled(InputElement).attrs<InputProps>({customKeys})<InputProps>(({theme,type$,fill$,disabled}) => `
    flex: 0 0 auto;
    display: none;

    + ${Foundation} {
        position: relative;
        flex: 0 0 auto;
        box-sizing: border-box;
        border-radius: 20px;
        border: solid 1px ${theme.src.type[type$].form.radioBorder};
        background: ${fill$ ? theme.src.type[type$].form.radioBorder : theme.src.type[type$].form.radioColor};
        box-shadow: ${theme.src.type[type$].form.radioBorder} 0 0 0 0 inset;
        transition: 400ms 0s ease-in-out;
        transition-property: border, box-shadow;

        > ${Slider} {
            position: absolute;
            left: 0;
            top: 0;
            border-radius: 100%;
            background: ${theme.src.type[type$].form.radioColor};
            box-shadow: 0 1px 3px rgba(0,0,0,.4);
            transition: left 200ms 0s ease-in-out;
        }
    }

    &:checked + ${Foundation} {
        border-color: ${theme.src.type[type$].main};
        background: ${theme.src.type[type$].main};
        box-shadow: ${theme.src.type[type$].main} 0 0 0 11px inset;
        transition: border 400ms 0s ease-in-out, box-shadow 400ms 0s ease-in-out, background 0s 400ms ease-in-out;
    }

    ${disabled ? `
        ~ * {
            opacity: ${theme.src.form.disabledOpacity};
            user-select: none;
        }
    ` : ""}
`)

export interface Props extends SizeProps {
    containerProps?: JSXElementProps
    inputProps?: JSXElementProps
    spanProps?: JSXElementProps
    children?: JSXChild
    type?: string
    fill?: boolean
    value?: Value
    active?: Value[]
    setActive?: SetValue<Value[]>
    checked?: boolean
    setChecked?: SetValue<boolean>
}

const Switch: React.FC<Props> = ({size="default",sizeSm,sizeMd,sizeLg,sizeXl,containerProps=DV.JSX_ELEMENT_PROPS,inputProps=DV.JSX_ELEMENT_PROPS,spanProps=DV.JSX_ELEMENT_PROPS,children,type="primary",fill=true,value,active,setActive,checked,setChecked}) => {
    const {checked$, change} = useMultiple({value,active,setActive,checked,setChecked,disabled:inputProps.disabled,onChange:inputProps.onChange})
    
    return (
        <RadioLabel {...containerProps} space={children ? true : false}>
            <Input {...inputProps} type="checkbox" size$={size} size$Sm={sizeSm} size$Md={sizeMd} size$Lg={sizeLg} size$Xl={sizeXl} type$={type} fill$={fill} checked={checked$} onChange={change} />
            <Foundation>
                <Slider />
            </Foundation>
            {children && <RadioSpan {...spanProps}>{children}</RadioSpan>}
        </RadioLabel>
    )
}

export default Switch