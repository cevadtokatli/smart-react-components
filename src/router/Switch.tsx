import React from "react"
import RouterContext from "./RouterContext"
import RouterHelper from "../helper/RouterHelper"

interface Props {
    children: JSX.Element[]
}

const Switch: React.FC<Props> = ({children}) => {
    const router = React.useContext(RouterContext)

    const getElements = () => {
        let switchNotRendered = false

        return children.map((item, index) => {
            let el = React.cloneElement(item, {key:(item.key || index),switchNotRendered})

            if(!switchNotRendered && RouterHelper.matchPath(router.state.url.pathname, router.state.url.query, {path:item.props.path,exact:item.props.exact,searchKeys:item.props.searchKeys}))
                switchNotRendered = true

            return el
        })
    }

    const [elements, setElements] = React.useState<JSX.Element[]>(() => getElements())
    const init = React.useRef<boolean>(false)

    React.useEffect(() => {
        if(!init.current)
            init.current = true
        else
            setElements(getElements())
    }, [children, router.state.url])

    return <>{elements}</>
}

export default Switch