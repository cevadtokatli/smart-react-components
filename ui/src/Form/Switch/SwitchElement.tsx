import Div from '@smart-react-components/core/Element/Div'
import { StyledProps } from '@smart-react-components/core/styled-props'
import { PaletteProp, ResponsiveProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import styled from 'styled-components'
import { extractNumberUnit, toCSSValue } from '../../util/css'

const Slider = styled.div(({ theme }) => `
  background: ${theme.$.color.dynamic.background};
  border-radius: 100%;
  box-shadow: 0 1px 3px rgba(0,0,0,.4);
  left: 0;
  position: absolute;
  top: 0;
  transition: 200ms 0s ease-in-out;
  transition-property: left, right;
`)

interface Props extends
  StyledProps,
  ResponsiveProp<'switchSize', SizeProp> {
  isChecked: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
}

export default styled(Div).attrs<Props>(({ isChecked }) => ({
  children: <Slider />,
  getSwitchSize: (v, t) => `
    &,
    + input {
      height: ${t.$.size.form.switch[v].height};
      width: ${t.$.size.form.switch[v].width};
    }

    + input {
      top: calc(50% - ${toCSSValue(t.$.size.form.switch[v].height)(v => v / 2)});
    }

    ${Slider} {
      height: ${t.$.size.form.switch[v].slider};
      width: ${t.$.size.form.switch[v].slider};

      ${isChecked
        ? `
          left: ${toCSSValue(t.$.size.form.switch[v].width)(width => width - extractNumberUnit(t.$.size.form.switch[v].slider) - 2)};  
        `
        : ''
      }
    }
  `,
}))<Props>(({ theme, isChecked, isOutline, isSoft, palette }: Props) => `
  border: solid 1px ${theme.$.color.dynamic.accent};
  border-radius: 20px;
  box-sizing: border-box;
  position: relative;
  transition: 400ms 0s ease-in-out;
  transition-property: border, box-shadow;

  ${!isOutline
    ? `
      background: ${theme.$.color.dynamic.accent};
      box-shadow: ${theme.$.color.dynamic.accent} 0 0 0 0 inset;
    `
    : ''
  }

  ${isChecked
    ? `
      transition: border 400ms 0s ease-in-out, box-shadow 400ms 0s ease-in-out, background 0s 400ms ease-in-out;

      ${!isSoft
        ? `
          background: ${theme.$.palette[palette].main};
          border-color: ${theme.$.palette[palette].main};
          box-shadow: ${theme.$.palette[palette].main} 0 0 0 11px inset;
        `
        : `
          background: ${theme.$.palette[palette].soft};
          border-color: ${theme.$.palette[palette].soft};
          box-shadow: ${theme.$.palette[palette].soft} 0 0 0 11px inset;
        `
      }
    `
    : ''
  }
`)
