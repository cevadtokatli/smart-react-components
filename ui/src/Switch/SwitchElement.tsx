import Div from '@smart-react-components/core/Element/Div'
import React from 'react'
import styled from 'styled-components'
import { Props } from '../Checkbox/CheckboxElement'
import { extractNumberUnit, toCSSValue } from '../util/css'

const Slider = styled.div(({ theme }) => `
  background: ${theme.$.color.dynamic.background};
  border-radius: 100%;
  box-shadow: 0 1px 3px rgba(0,0,0,.4);
  left: 0;
  position: absolute;
  top: 0;
  transition: 200ms 0s ease-in-out;
  transition-property: left, right;
`)

export default styled(Div).attrs<Props>(({ children, isChecked }) => ({
  children: (
    <>
      {children}
      <Slider />
    </>
  ),
  getCheckboxSize: (v, t) => `
    height: ${t.$.size.switch[v].height};
    width: ${t.$.size.switch[v].width};

    ${Slider} {
      height: ${t.$.size.switch[v].slider};
      width: ${t.$.size.switch[v].slider};

      ${isChecked
        ? `
          left: ${toCSSValue(t.$.size.switch[v].width)(width => width - extractNumberUnit(t.$.size.switch[v].slider) - 2)};  
        `
        : ''
      }
    }
  `,
}))<Props>(({ theme, isChecked, isOutline, isSoft, palette }: Props) => `
  border: solid 1px ${theme.$.color.dynamic.accent};
  border-radius: 20px;
  box-sizing: border-box;
  position: relative;
  transition: 400ms 0s ease-in-out;
  transition-property: border, box-shadow;

  ${!isOutline
    ? `
      background: ${theme.$.color.dynamic.accent};
      box-shadow: ${theme.$.color.dynamic.accent} 0 0 0 0 inset;
    `
    : ''
  }

  ${isChecked
    ? `
      transition: border 400ms 0s ease-in-out, box-shadow 400ms 0s ease-in-out, background 0s 400ms ease-in-out;

      ${!isSoft
        ? `
          background: ${theme.$.palette[palette].main};
          border-color: ${theme.$.palette[palette].main};
          box-shadow: ${theme.$.palette[palette].main} 0 0 0 11px inset;
        `
        : `
          background: ${theme.$.palette[palette].soft};
          border-color: ${theme.$.palette[palette].soft};
          box-shadow: ${theme.$.palette[palette].soft} 0 0 0 11px inset;
        `
      }
    `
    : ''
  }
`)
