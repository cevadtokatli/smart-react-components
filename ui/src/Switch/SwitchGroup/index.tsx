import React from 'react'
import CheckboxGroup, { Props as CheckboxGroupProps } from '../../Checkbox/CheckboxGroup'

export { OrderPosition as Position } from '../../types'

export type Props = Omit<CheckboxGroupProps, 'shape'>

const SwitchGroup: React.FC<Props> = CheckboxGroup

export default SwitchGroup
