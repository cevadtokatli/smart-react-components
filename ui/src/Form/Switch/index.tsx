import React from 'react'
import { OrderPosition } from '../../types'
import Checkbox, { Props as CheckboxProps } from '../Checkbox'
import SwitchElement from './SwitchElement'

export { OrderPosition as Position } from '../../types'

export type Props = Omit<CheckboxProps, 'shape'>

const Switch = React.forwardRef<HTMLInputElement, Props>((props, forwardRef) => (
  <Checkbox
    {...props}
    render={props.render ?? (({ isChecked }) => (
      <SwitchElement
        isChecked={isChecked}
        isOutline={props.isOutline}
        isSoft={props.isSoft}
        palette={props.palette}
        switchSize={props.size}
        switchSizeSm={props.sizeSm}
        switchSizeMd={props.sizeMd}
        switchSizeLg={props.sizeLg}
        switchSizeXl={props.sizeXl}
      />)
    )}
    ref={forwardRef}
  />
))

Switch.defaultProps = {
  palette: 'primary',
  position: OrderPosition.RIGHT,
  size: 'medium',
}

export default Switch
