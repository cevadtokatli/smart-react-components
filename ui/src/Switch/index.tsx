import React from 'react'
import Checkbox, { Props as CheckboxProps } from '../Checkbox'
import SwitchElement from '../components/Switch/SwitchElement'

export { OrderPosition as Position } from '../types'

export type Props = Omit<CheckboxProps, 'shape'>

const Switch = React.forwardRef<HTMLInputElement, Props>((props, forwardRef) => <Checkbox {...props} ref={forwardRef} />)

Switch.defaultProps = {
  ...Checkbox.defaultProps,
  template: <SwitchElement />,
}

export default Switch
