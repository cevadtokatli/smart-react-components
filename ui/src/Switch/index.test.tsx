import { wrapTheme } from '@smart-react-components/core/test'
import { render } from '@testing-library/react'
import React from 'react'
import Switch from './'

describe('<Switch />', () => {
  it('should render component', () => {
    const { asFragment } = render(wrapTheme(
      <Switch value="value" active={['value']} setActive={jest.fn()} label="Label" />
    ))
    expect(asFragment()).toMatchSnapshot()
  })
})
