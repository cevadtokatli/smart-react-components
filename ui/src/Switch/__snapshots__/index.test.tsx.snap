// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> should render component 1`] = `
<DocumentFragment>
  .c0 {
  margin: 1rem .5rem;
}

.c0 > span {
  font-size: 1rem;
  margin-left: 0.46874rem;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: default;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
}

.c1 > div {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c1 > span {
  color: #323a46;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  font-family: inherit;
  word-break: break-word;
  word-wrap: break-word;
  -webkit-tap-highlight-color: transparent;
}

.c4 {
  height: 100%;
  left: 0;
  opacity: 0;
  pointer-events: none;
  position: absolute;
  top: 0;
  width: 100%;
}

.c2 {
  height: 16px;
  width: 28px;
}

.c2 .c5 {
  height: 14px;
  width: 14px;
  left: 12px;
}

.c6 {
  background: #fff;
  border-radius: 100%;
  box-shadow: 0 1px 3px rgba(0,0,0,.4);
  left: 0;
  position: absolute;
  top: 0;
  -webkit-transition: 200ms 0s ease-in-out;
  transition: 200ms 0s ease-in-out;
  -webkit-transition-property: left,right;
  transition-property: left,right;
}

.c3 {
  border: solid 1px #dbdbdb;
  border-radius: 20px;
  box-sizing: border-box;
  position: relative;
  -webkit-transition: 400ms 0s ease-in-out;
  transition: 400ms 0s ease-in-out;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  background: #dbdbdb;
  box-shadow: #dbdbdb 0 0 0 0 inset;
  -webkit-transition: border 400ms 0s ease-in-out,box-shadow 400ms 0s ease-in-out,background 0s 400ms ease-in-out;
  transition: border 400ms 0s ease-in-out,box-shadow 400ms 0s ease-in-out,background 0s 400ms ease-in-out;
  background: #0665d0;
  border-color: #0665d0;
  box-shadow: #0665d0 0 0 0 11px inset;
}

<label
    class="c0 c1"
  >
    <div
      class="c2 c3"
      shape="rectangle"
    >
      <input
        checked=""
        class="c4"
        type="checkbox"
      />
      <div
        class="c5 c6"
      />
    </div>
    <span>
      Label
    </span>
  </label>
</DocumentFragment>
`;
